### DISCUSSION TITLE
How do we want to handle releases

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#84] [Development Practices]
Traditionally @mikecovesw has create releases has various users have asked for features. Given that we're more centralized and we have a handful of devs and fairly continuous additions and improvements we should come up with a release method/cycle.

My thought is monthly and make it clear if something is a security release (for those of us stuck with certain gov't rules.)

Or we can just put things up to a vote every so often.

Any other ideas?

We've gotten the build structured (either in master or 7.1) where anyone can generate a release and we seem to have gotten things able to be in 1 install jar. So other than all the regression tests (which I hope to get in the automated workflow soon but we can run manually) any one can do the process of building and uploading to github.

**Comment by agilmore2** [comment]
Release pace should mirror development pace. If we're making enough changes that monthly releases make sense, let's do it. Otherwise, something like the fedora karma system where regular builds get votes and at some threshold they're released makes sense to me.

**Reply by MikeNeilson** [reply]
That's a good point. Plus we have the government users who's IT departments might push annoying update policies with out real need.

Perhaps the following:
- when one of us things there's enough changes, or there's a very specific need by end users we pitch a vote.
  - we would then need to all commit to at least some about of testing and verification, especially until we've automated the regression tests available.
- Regardless of development pace we do no more than monthly.
  - exceptions, of course, for security issues and major usability problems.

As far as released version go. I've certainly pitched some drastic changes to ease future development that while they shouldn't affect general behavior would be disruptive until a substantial amount of change is finished.

We currently have the Major.Minor.RC release numbering. 
We'll likely want to commit to keeping 

| branch/version | definition | workload | 
|------------------|------------|-----------|
| main | all devs contribute here| not specifically released to users. Jars and builds available for downstream testing and experimentation.|
|x.y.z | Current development released code| Bug fixes, Behavioral changes; basically tracks main. Various manual and automated regression tests have been done.|
|x.y-1.z| If x.y.z altered enough but similar| Bug fixes, Behavioral changes *IFF* not difficult|
|x-1.y.z| Transtition support. 1 year max| Bug fixes. no behavior changes; example 6.8 to 7.0 - we're not backporting the connection pool idea as it would be impossible|

granted x.y-1 and x-1 may or may not exist at any given point; but going forward as we identify larger changes it seems appropriate not to immediately force everyone through a loop but maintain a balance of exactly how much code we're maintaining.

For example, I'm currently keeping the 7.1 branch up-to-date against master. My thought is once I've gotten the build complete we copy 7.1 to main (making "main" the default branch as is the newer standard for default branch names) and move the current master to 7.0. And those would be the supported versions.

Though the only really change I made in 7.1 was updating the CWMS DB dependencies and not requiring the database jars in a different classpath/folder so arguably it doesn't doesn't justify 7.1; mostly did that because I didn't want to accidentally mix up something.

**Comment by agilmore2** [comment]
Semantic versioning is a good thing in my opinion. Since Precision is paid
to support HDB and OpenDCS, we're going to support our users on whatever
version, and having a policy like this would help consolidate their current
usage.

I'm in favor.

Andrew

**Comment by MikeNeilson** [comment]
While we're certainly still having issues actually implementing the release plan. I think we all more or less agree. Should anyone want to propose changes to how we operate please open a new discussion.
